### PR TITLE
Refactoring of Add Task List and Add Task Item

### PR DESCRIPTION
**Pull request by kumarpalsinh25**
Ref #1808

Reference VIdeo:

https://github.com/acterglobal/a3/assets/51526480/573ec1f8-3d48-47b0-8df4-bc31c6962326

### CODE REVIEW

File: app/lib/features/tasks/widgets/task_list_card.dart [hunk 1]
```diff
@@ -134,19 +138,17 @@ class _TaskListCardState extends ConsumerState<TaskListCard> {
...
                     children.add(
-                      Padding(
+                      Container(
+                        alignment: Alignment.centerLeft,
                         padding: const EdgeInsets.symmetric(
                           horizontal: 24,
                           vertical: 8,
                         ),
-                        child: OutlinedButton(
+                        child: ActerInlineTextButton(
```
Comment (gnunicorn) [inline app/lib/features/tasks/widgets/task_list_card.dart:147]: :+1:

File: app/lib/features/tasks/sheets/create_update_task_item.dart [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            hintText: L10n.of(context).dueDate,
+            suffixIcon: IconButton(
+              onPressed: selectDueDate,
+              icon: const Icon(Icons.calendar_month),
+            ),
+          ),
+          onTap: selectDueDate,
+          controller: _taskDueDateController,
+        ),
+      ],
```
Comment (gnunicorn) [inline app/lib/features/tasks/sheets/create_update_task_item.dart:173]: when changing the tasks due date the most common way to think about it that something is due "today" or "tomorrow" (and patterns alike). That's why for changing the due-date we have a custom date picker that allows for those to easily be picked. I understand the want to not reuse that (rather ugly) widget but we need to keep that use case in mind. Maybe by adding two `ActerInlineTextButtons` for quick setting? (the following is not real applicable code, just to translate an idea)
```suggestion
        TextFormField(
          readOnly: true,
          decoration: InputDecoration(
            hintText: L10n.of(context).dueDate,
            suffixIcon: IconButton(
              onPressed: selectDueDate,
              icon: const Icon(Icons.calendar_month),
            ),
          ),
          onTap: selectDueDate,
          controller: _taskDueDateController,
        ),
        Row(children:[
          ActerInlineTextButton(child: Text(L10n.of(context).today, onPressed: () => setState(() => { selectedDate = DateTime.today() }), 
           ActerInlineTextButton(child: Text(L10n.of(context).tomorrow, onPressed: () => setState(() => { selectedDate = DateTime.tomorrow() }),
           ])
      ],
```

File: app/lib/features/tasks/sheets/create_update_task_item.dart [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    final date = await showDatePicker(
+      context: context,
+      initialDate: selectedDate,
+      firstDate: DateTime.now(),
+      lastDate: DateTime.now().addYears(1),
+    );
+    if (date == null || !mounted) return;
+    selectedDate = date;
+    _taskDueDateController.text = taskDueDateFormat(date);
+    setState(() {});
```
Comment (gnunicorn) [inline app/lib/features/tasks/sheets/create_update_task_item.dart:187]: seen this empty set-state pattern a few times, but shouldn't this rather be:
```suggestion
    setState(() {
      selectedDate = date;
      _taskDueDateController.text = taskDueDateFormat(date);
    });
```
?

File: app/lib/features/tasks/sheets/create_update_task_item.dart [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    final taskDraft = widget.taskList.taskBuilder();
+    taskDraft.title(_taskNameController.text);
+    if (_taskDescriptionController.text.isNotEmpty) {
+      taskDraft.descriptionText(_taskDescriptionController.text);
+    }
+    if (selectedDate != null) {
+      taskDraft.dueDate(
+        selectedDate!.year,
+        selectedDate!.month,
+        selectedDate!.day,
```
Comment (gnunicorn) [inline app/lib/features/tasks/sheets/create_update_task_item.dart:209]: these `!` shouldn't be necessary due to your check above.
```suggestion
        selectedDate.year,
        selectedDate.month,
        selectedDate.day,
```

File: app/lib/features/tasks/sheets/create_update_task_item.dart [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+      );
+    }
+    try {
+      await taskDraft.send();
+      EasyLoading.dismiss();
+      if (!mounted) return;
+      if (widget.cancel != null) widget.cancel!();
```
Comment (gnunicorn) [inline app/lib/features/tasks/sheets/create_update_task_item.dart:216]: why are you calling `cancel` on a successful submission? that seems unexpected...

File: app/lib/features/tasks/sheets/create_update_task_list.dart [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+            Text(
+              L10n.of(context).description,
+              style: Theme.of(context).textTheme.bodySmall,
+            ),
+            MdEditorWithPreview(
+              key: CreateUpdateTaskList.descKey,
+              onChanged: (String? value) {
+                _descriptionText.value = value ?? '';
+              },
+            ),
```
Comment (gnunicorn) [inline app/lib/features/tasks/sheets/create_update_task_list.dart:140]: we are phasing out the MDEditorWithPreview in favor of the AppFlowyEditor. We should use that one here instead.

File: app/lib/features/tasks/sheets/create_update_task_item.dart [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+        ),
+        const SizedBox(height: 5),
+        TextFormField(
+          decoration: InputDecoration(
+            hintText: L10n.of(context).description,
+          ),
+          minLines: 4,
+          maxLines: 4,
+          controller: _taskDescriptionController,
+        ),
```
Comment (gnunicorn) [inline app/lib/features/tasks/sheets/create_update_task_item.dart:147]: isn't this a fully markup supporting field? We should use the appflowy editor here...

**Review by gnunicorn — APPROVED** [review]
Great work. Just a few usability improvements

File: app/lib/features/tasks/sheets/create_update_task_item.dart (hunk 2 above)
Comment (kumarpalsinh25) [reply app/lib/features/tasks/sheets/create_update_task_item.dart:173]: Added options as below.

Screenshot 2024-06-05 at 6 35 29 PM

File: app/lib/features/tasks/sheets/create_update_task_item.dart (hunk 4 above)
Comment (kumarpalsinh25) [reply app/lib/features/tasks/sheets/create_update_task_item.dart:209]: Need to keep it as it variable is global and giving syntax error.

Screenshot 2024-06-05 at 6 36 50 PM

File: app/lib/features/tasks/sheets/create_update_task_item.dart (hunk 5 above)
Comment (kumarpalsinh25) [reply app/lib/features/tasks/sheets/create_update_task_item.dart:216]: This is to give call back to outer screen and close inline input box.

File: app/lib/features/tasks/sheets/create_update_task_list.dart (hunk 6 above)
Comment (kumarpalsinh25) [reply app/lib/features/tasks/sheets/create_update_task_list.dart:140]: Can we do this on separate PR as I first I would like to focus on Refactoring task UI-UX to quickly get out of lab feature and then improve individual smaller pieces on quick PRs.

Side-note:
- Not sure but probably also need to replace `taskListDraft.descriptionMarkdown` with `taskListDraft.descriptionHtml(plaintext, htmltext)`.

File: app/lib/features/tasks/sheets/create_update_task_item.dart (hunk 7 above)
Comment (kumarpalsinh25) [reply app/lib/features/tasks/sheets/create_update_task_item.dart:147]: Same as above.

File: app/lib/features/tasks/sheets/create_update_task_item.dart (hunk 4 above)
Comment (gnunicorn) [reply app/lib/features/tasks/sheets/create_update_task_item.dart:230]: odd. looks like a false-positive of your IDE

**Pull request merged**